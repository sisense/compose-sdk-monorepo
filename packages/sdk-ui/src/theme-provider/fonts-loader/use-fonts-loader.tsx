/* eslint-disable sonarjs/no-nested-template-literals */
import { ThemeSettingsFont } from '@/types';
import { useEffect, useMemo, useRef, useState } from 'react';

export const useFontsLoader = (fonts: ThemeSettingsFont[], prevLoadedFonts: string[]) => {
  const styleElement = useRef(document.createElement('style'));
  const [loadedFonts, setLoadedFonts] = useState<string[]>([]);

  const notLoadedFonts = useMemo(
    () => fonts.filter((font) => !prevLoadedFonts.includes(getFontInditifier(font))),
    [fonts, prevLoadedFonts],
  );

  useEffect(() => {
    if (notLoadedFonts.length) {
      styleElement.current = document.createElement('style');
      const fontFaces = notLoadedFonts.map(prepareFontFaceString);
      styleElement.current.innerHTML = fontFaces.join('\n');

      document.head.appendChild(styleElement.current);
      setLoadedFonts(notLoadedFonts.map(getFontInditifier));
    }

    return () => {
      styleElement.current?.remove();
    };
  }, [notLoadedFonts]);

  return loadedFonts;
};

function getFontInditifier(font: ThemeSettingsFont) {
  return `${font.fontFamily}|${font.fontWeight}|${font.fontStyle}`;
}

function prepareFontFaceString(font: ThemeSettingsFont) {
  return `@font-face {
          font-family: "${font.fontFamily}";
          font-weight: ${font.fontWeight};
          font-style: ${font.fontStyle};
          ${font.src
            .map(
              (src) =>
                `src: ${Object.entries(src)
                  .map(([key, value]) => `${key}('${value}')`)
                  .join(' ')};`,
            )
            .join('\n')}
      }`;
}
