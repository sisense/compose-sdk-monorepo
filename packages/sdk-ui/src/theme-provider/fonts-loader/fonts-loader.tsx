import { ThemeSettingsFont } from '@/types';
import { useFontsLoader } from '@/theme-provider/fonts-loader/use-fonts-loader';
import { useThemeContext } from '@/theme-provider';
import { createContext, useContext } from 'react';

const FontLoaderContext = createContext<{
  loadedFonts: string[];
}>({
  loadedFonts: [],
});

export const FontsLoader = ({
  fonts,
  children,
}: {
  fonts?: ThemeSettingsFont[];
  children?: any;
}) => {
  const { loadedFonts: prevLoadedFonts } = useContext(FontLoaderContext);
  const { themeSettings } = useThemeContext();
  const { loadedFonts, areAllFontsLoaded } = useFontsLoader(
    fonts || themeSettings.typography.fontsLoader?.fonts || [],
    prevLoadedFonts,
  );

  return (
    <FontLoaderContext.Provider value={{ loadedFonts: [...prevLoadedFonts, ...loadedFonts] }}>
      {areAllFontsLoaded && children}
    </FontLoaderContext.Provider>
  );
};
