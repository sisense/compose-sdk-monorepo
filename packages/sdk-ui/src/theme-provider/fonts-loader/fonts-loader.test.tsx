import { describe } from 'vitest';
import { FontsLoader } from '@/theme-provider/fonts-loader/fonts-loader';
import { render, waitFor } from '@testing-library/react';

const FONT_MOCK = {
  fontFamily: 'FontTest1',
  fontWeight: 'bold',
  fontStyle: 'normal',
  src: [
    {
      url: `FontTest1-Bold.eot`,
    },
    {
      local: `FontTest1-Bold`,
    },
    {
      url: `FontTest1-Bold.eot?#iefix`,
      format: 'embedded-opentype',
    },
    {
      url: `FontTest1-Bold.ttf`,
      format: 'truetype',
    },
  ],
};

describe('FontLoader', () => {
  it('should load font correctly', async () => {
    render(<FontsLoader fonts={[FONT_MOCK]}>Test</FontsLoader>);

    await waitFor(() => {
      expect(document.querySelector('head')).toMatchSnapshot();
    });
  });

  it('should avoid duplication on multiple instance', async () => {
    render(
      <FontsLoader fonts={[FONT_MOCK]}>
        <FontsLoader fonts={[FONT_MOCK]}>Test</FontsLoader>
      </FontsLoader>,
    );

    await waitFor(() => {
      expect(document.querySelector('head')).toMatchSnapshot();
    });
  });

  it('should load different fonts on multiple instance', async () => {
    render(
      <FontsLoader fonts={[FONT_MOCK]}>
        <FontsLoader fonts={[{ ...FONT_MOCK, fontFamily: 'AnotherFont' }]}>Test</FontsLoader>
      </FontsLoader>,
    );

    await waitFor(() => {
      expect(document.querySelector('head')).toMatchSnapshot();
    });
  });
});
