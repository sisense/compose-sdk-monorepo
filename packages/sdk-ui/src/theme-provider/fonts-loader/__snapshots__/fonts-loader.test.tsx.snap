// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FontLoader > should avoid duplication on multiple instance 1`] = `
<head>
  <style>
    @font-face {
          font-family: "FontTest1";
          font-weight: bold;
          font-style: normal;
          src: url('FontTest1-Bold.eot');
src: local('FontTest1-Bold');
src: url('FontTest1-Bold.eot?#iefix') format('embedded-opentype');
src: url('FontTest1-Bold.ttf') format('truetype');
      }
  </style>
</head>
`;

exports[`FontLoader > should load different fonts on multiple instance 1`] = `
<head>
  <style>
    @font-face {
          font-family: "FontTest1";
          font-weight: bold;
          font-style: normal;
          src: url('FontTest1-Bold.eot');
src: local('FontTest1-Bold');
src: url('FontTest1-Bold.eot?#iefix') format('embedded-opentype');
src: url('FontTest1-Bold.ttf') format('truetype');
      }
  </style>
  <style>
    @font-face {
          font-family: "AnotherFont";
          font-weight: bold;
          font-style: normal;
          src: url('FontTest1-Bold.eot');
src: local('FontTest1-Bold');
src: url('FontTest1-Bold.eot?#iefix') format('embedded-opentype');
src: url('FontTest1-Bold.ttf') format('truetype');
      }
  </style>
</head>
`;

exports[`FontLoader > should load font correctly 1`] = `
<head>
  <style>
    @font-face {
          font-family: "FontTest1";
          font-weight: bold;
          font-style: normal;
          src: url('FontTest1-Bold.eot');
src: local('FontTest1-Bold');
src: url('FontTest1-Bold.eot?#iefix') format('embedded-opentype');
src: url('FontTest1-Bold.ttf') format('truetype');
      }
  </style>
</head>
`;
